src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.Analysis.cs

```diff
@@ -0,0 +1,98 @@
+
+        private void AnalyzeIfEnabled(SemanticModelAnalysisContext context, Action<SemanticModelAnalysisContext, CodeStyleOption<TOptionKind>> analyze)
+            => AnalyzeIfEnabled(context, analyze, context.Options, context.SemanticModel.SyntaxTree, context.CancellationToken);
+
+        private void AnalyzeIfEnabled(SyntaxNodeAnalysisContext context, Action<SyntaxNodeAnalysisContext, CodeStyleOption<TOptionKind>> analyze)
+            => AnalyzeIfEnabled(context, analyze, context.Options, context.SemanticModel.SyntaxTree, context.CancellationToken);
+
+        private void AnalyzeIfEnabled(SyntaxTreeAnalysisContext context, Action<SyntaxTreeAnalysisContext, CodeStyleOption<TOptionKind>> analyze)
+            => AnalyzeIfEnabled(context, analyze, context.Options, context.Tree, context.CancellationToken);
+
+        private void AnalyzeIfEnabled<TContext>(
+            TContext context, Action<TContext, CodeStyleOption<TOptionKind>> analyze,
+            AnalyzerOptions options, SyntaxTree syntaxTree, CancellationToken cancellationToken)
+        {
+            var optionSet = options.GetDocumentOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult();
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var optionValue = optionSet.GetOption(_option);
+            switch (optionValue.Notification.Severity)
+            {
+                case ReportDiagnostic.Error:
+                case ReportDiagnostic.Warn:
+                case ReportDiagnostic.Info:
+                    break;
+                default:
+                    // don't analyze if it's any other value.
+                    return;
+            }
+
+            analyze(context, optionValue);
+        }
+
+        public abstract class DiagnosticAnalyzer : AbstractCodeStyleDiagnosticAnalyzer
+        {
+            public readonly TCodeStyleProvider _codeStyleProvider;
+
+            protected DiagnosticAnalyzer(bool configurable = true) 
+                : this(new TCodeStyleProvider(), configurable)
+            {
+            }
+
+            private DiagnosticAnalyzer(TCodeStyleProvider codeStyleProvider, bool configurable)
+                : base(codeStyleProvider._descriptorId, codeStyleProvider._title, codeStyleProvider._message, configurable)
+            {
+                _codeStyleProvider = codeStyleProvider;
+            }
+
+            protected sealed override void InitializeWorker(AnalysisContext context)
+                => _codeStyleProvider.DiagnosticAnalyzerInitialize(context);
+
+            public sealed override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+                => _codeStyleProvider.GetDiagnosticAnalyzerCategory();
+
+            public sealed override bool OpenFileOnly(Workspace workspace)
+                => _codeStyleProvider.DiagnosticsForOpenFileOnly(workspace);
```

Comment:
all members in these nested classes are intentionally sealed. All logic should be provided in the CodeStyleProvider subclass.  The feature classes that subclass these nested classes should nto be able to override these themselves.